src/atheros/atherosd.c

```diff
@@ -413,51 +351,45 @@ void *serial_monitor(void *arg)
     uint8_t messageBuffer[64];
     uint8_t messageBufferIndex = 0;
 
-    uint8_t magicBytes[] = {0x47, 0x41, 0x4e, 0x53};
     uint8_t messageLengthArray[] = {9, 6};
 
     while(1)
     {
-        pthread_mutex_lock(&state->atmegafp_mutex);
-        int data = getc(state->atmegafp_read);
-        pthread_mutex_unlock(&state->atmegafp_mutex);
+        int8_t data = fgetc(state->atmegafp);
 
-        if(data != EOF)
-        {
-            messageBuffer[messageBufferIndex] = data;
+        messageBuffer[messageBufferIndex] = data;
 
-            if(messageBufferIndex <= 3 && messageBuffer[messageBufferIndex] != magicBytes[messageBufferIndex])
-            {
-                memset(messageBuffer, 0, messageBufferIndex + 1);
-                messageBufferIndex = 0;
-            }
+        if(messageBufferIndex <= 3 && messageBuffer[messageBufferIndex] != magic_bytes[messageBufferIndex])
+        {
+            memset(messageBuffer, 0, messageBufferIndex + 1);
```

Comment:
I think you should be more liberal with you memsets. Each time you are moving on to a new message, you want to flush the entire buffer, right? Just memset the whole thing, even if you don't think it matters. Also, you should be sure to memset the buffer before use. Until you do so, it just contains garbage. Being really diligent about these kinds of memory things help prevent nasty memory corruption bugs.